packages/react/src/template.tsx

```diff
@@ -117,36 +130,49 @@ export function BaseTemplate({
   dataState: DataState
   onTemplateEvent?: (event: TemplateEvent) => any
 }) {
-  function renderComponent(stateNode: ComponentDataState, path: Path) {
-    const componentDefinition = components[stateNode.component]
-    if (!componentDefinition) {
-      throw new RenderError(`Unknown component: ${stateNode.component}`)
-    }
-
-    // tbd: validate `components` prop once with `zod` instead of later render stage
-    const Component = componentDefinition.component
-    if (!Component) {
-      throw new RenderError(`Invalid component: ${stateNode.component}`)
-    }
+  const RenderComponent = useCallback(
+    function ({ stateNode, path }: { stateNode: ComponentDataState; path: Path }) {
+      const componentDefinition = components[stateNode.component]
+      if (!componentDefinition) {
+        throw new RenderError(`Unknown component: ${stateNode.component}`)
+      }
 
-    let componentProps = Object.fromEntries(
-      Object.entries(stateNode.props || {}).map(([propKey, propValue]) => {
-        return [propKey, renderProp(propKey, propValue, stateNode, [...path, 'props', propKey])]
-      })
-    )
-    if (typeof componentDefinition.validator === 'function') {
-      try {
-        componentProps = componentDefinition.validator(componentProps)
-      } catch (e) {
-        throw new RenderError(
-          `Invalid props for component: ${stateNode.component}, props: ${JSON.stringify(stateNode.props)}. Error: ${JSON.stringify(e)}`
-        )
+      const Component = componentDefinition.component
+      if (!Component) {
+        throw new RenderError(`Invalid component: ${stateNode.component}`)
       }
-    }
 
-    const children = stateNode.children ? render(stateNode.children, [...path, 'children']) : null
+      const localState = useContext(LocalState)
+      let componentProps = Object.fromEntries(
+        Object.entries(stateNode.props || {}).map(([propKey, propValue]) => {
+          return [
+            propKey,
+            renderProp(propKey, propValue, stateNode, localState, [...path, 'props', propKey]),
+          ]
+        })
+      )
+      if (typeof componentDefinition.validator === 'function') {
+        try {
+          componentProps = componentDefinition.validator(componentProps)
+        } catch (e) {
+          throw new RenderError(
```

Comment:
unrelated to this change but we probably want to pass in the validation error as a first-class value so that the suspense boundary can have nice logic for addressing the error

Reply:
will turn this into issue - haven't thought much about error handling, so don't know how to answer this yet, but it's important for sure.

Reply:
Yeah definitely important pre v1 release, but maybe we don't have to worry for the beta (release for the first conf)